### PR TITLE
perf(spec-parser): update list function to support invalid reason

### PR DESCRIPTION
**Pull request by SLdragon**
Update list function response type:

```ts
export interface ListAPIResult {
  allAPICount: number;
  validAPICount: number;
  APIs: ListAPIInfo[];
}

export interface ListAPIInfo {
  api: string;
  server: string;
  operationId: string;
  isValid: boolean;
  reason: ErrorType[];
  auth?: AuthInfo;
}
```

### CODE REVIEW

**Review by yuqizhou77 — DISMISSED** [review]

File: packages/spec-parser/src/specFilter.ts [hunk 1]
```diff
@@ -22,24 +22,32 @@
...
+          }
 
-        if (!newPaths[path]) {
-          newPaths[path] = { ...unResolveSpec.paths[path] };
-          for (const m of ConstantString.AllOperationMethods) {
-            delete (newPaths[path] as any)[m];
+          if (!newPaths[path]) {
+            newPaths[path] = { ...unResolveSpec.paths[path] };
+            for (const m of ConstantString.AllOperationMethods) {
+              delete (newPaths[path] as any)[m];
```
[the remark this answers is not in the document]
Comment (SLdragon) [reply packages/spec-parser/src/specFilter.ts:39]: Despite the prototype pollution warning, our code includes checks within the loop to ensure that m/methodName is one of the supported operation methods. This effectively mitigates the risk of prototype pollution, making the warning unnecessary in this context.

File: packages/spec-parser/src/specFilter.ts [hunk 2]
```diff
@@ -22,24 +22,32 @@
...
+            }
           }
-        }
 
-        (newPaths[path] as any)[methodName] = (unResolveSpec.paths[path] as any)[methodName];
+          (newPaths[path] as any)[methodName] = (unResolveSpec.paths[path] as any)[methodName];
```
[the remark this answers is not in the document]
Comment (SLdragon) [reply packages/spec-parser/src/specFilter.ts:43]: Despite the prototype pollution warning, our code includes checks within the loop to ensure that m/methodName is one of the supported operation methods. This effectively mitigates the risk of prototype pollution, making the warning unnecessary in this context.

**Review by yuqizhou77 — APPROVED** [review]

File: packages/spec-parser/src/utils.ts [hunk 3]
```diff
@@ -36,10 +38,11 @@ export class Utils {
     paramObject: OpenAPIV3.ParameterObject[],
     isCopilot: boolean
```
Comment (adashen) [inline packages/spec-parser/src/utils.ts:39]: the method is shared for api me, type b, custom copilot?

File: packages/spec-parser/src/utils.ts [hunk 4]
```diff
@@ -179,119 +189,147 @@ export class Utils {
     path: string,
     spec: OpenAPIV3.Document,
     options: ParseOptions
-  ): boolean {
-    const pathObj = spec.paths[path] as any;
+  ): APIValidationResult {
```
Comment (adashen) [inline packages/spec-parser/src/utils.ts:192]: This method will also check validation for custom copilot scenario?

File: packages/spec-parser/src/utils.ts (hunk 3 above)
Comment (SLdragon) [reply packages/spec-parser/src/utils.ts:39]: Yes, in the future, will separate logic for different project type

File: packages/spec-parser/src/utils.ts (hunk 4 above)
Comment (SLdragon) [reply packages/spec-parser/src/utils.ts:192]: Yes, same above, currently for all

**Review by xzf0587 — APPROVED** [review]

**Review by nliu-ms — APPROVED** [review]

**Review by adashen — APPROVED** [review]

**Review by jayzhang — APPROVED** [review]

**Comment by KennethBWSong** [comment]
@SLdragon Also need to update [this](https://github.com/OfficeDev/TeamsFx/blob/a277dba4e949fe6276412a1dae0825f3f67c3521/packages/fx-core/src/component/generator/copilotPlugin/helper.ts#L219) and [this](https://github.com/OfficeDev/TeamsFx/blob/a277dba4e949fe6276412a1dae0825f3f67c3521/packages/fx-core/src/component/generator/copilotPlugin/helper.ts#L359)?

**Pull request merged**